c7n/resources/ec2.py

```diff
@@ -295,6 +296,57 @@ def __call__(self, i):
         return self.operator(map(self.match, volumes))
 
 
+@filters.register('stop-protected')
+class DisableApiStop(Filter):
+    """EC2 instances with ``disableApiStop`` attribute set
+
+    Filters EC2 instances with ``disableApiStop`` attribute set to true.
+
+    :Example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: stop-protection-enabled
+            resource: ec2
+            filters:
+              - type: stop-protected
+
+    :Example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: stop-protection-NOT-enabled
+            resource: ec2
+            filters:
+              - not:
+                - type: stop-protected
+    """
+
+    schema = type_schema('stop-protected')
+    permissions = ('ec2:DescribeInstanceAttribute',)
+
+    def get_permissions(self) -> List[str]:
+        perms = list(self.permissions)
+        perms.extend(self.manager.get_permissions())
+        return perms
+
+    def process(self, resources: List[dict], event=None) -> List[dict]:
+        client = utils.local_session(
+            self.manager.session_factory).client('ec2')
+        return [r for r in resources
+                if self._is_stop_protection_enabled(client, r)]
+
+    def _is_stop_protection_enabled(self, client, instance: dict) -> bool:
+        attr_val = self.manager.retry(
+            client.describe_instance_attribute,
+            Attribute='disableApiStop',
+            InstanceId=instance['InstanceId']
+        )
+        return attr_val['DisableApiStop']['Value']
```

Comment:
This is going to raise a `KeyError` for versions of botocore prior to 1.26.7 (the default shipping w/Lambda today is 1.23.32). Not sure the best way around that. The most explicit way to handle that might be letting the policy die, but also emitting a warning with some guidance around upgrading botocore :thinking: .

Another lower friction (but potentially misleading) option would be to default to false, emit a warning (that people may not read) and move on.

Reply:
we could also conditionally register the action based on the version of botocore.

Reply:
or fail in validation if min version isn't met

Reply:
Hi all,

Thanks for your comments and ideas.

Decided to implement comparing _botocore_ version in _validation_ method. To my mind, it's the most clear place (yet, not ideal) when the policy has to fail.

Quick question: have we got a way to (unit) test policy on Lambda mode (not found any info)?

Reply:
Hmm, I don't think the execution mode is particularly relevant for the check, I was just mentioning it because the default Lambda runtime represents the environment used by a good portion of actively deployed policies.

Validation check on the botocore version seems like a smart plan though, good call :+1: . For unit testing purposes it may be simplest to patch `botocore.__version__` to trigger the validation check.

Reply:
(Hmm, looks like you sorted out the unit test bits while I had the comment window open :sweat_smile: )